Traders: Tessera Standalone Script (almanac code based)

SekmehtDR opened the pull request:
- refactored almanac.lic script to `ask tessera about invest` every 6 min 15 sec.
- uses settings.almanac_no_use_scripts to avoid scripts being used with almanac too. (tessera_no_use_scripts could be added to base.yaml if really important).
- Been using it for moths, works great.

SekmehtDR commented:
Let me know if there are any concerns with this, at one point I ported over the almanac code and rewrote it for the Tessera. Thanks all.

rpherbig reviewed (changes requested):
Please take a look at the current version of `almanac.lic` as it has some improvements that should really be copied over to here.

For example, instead of using `pause_all` and `unpause_all`, which could pause critical script or unpause those which weren't paused by _this_ script, `almanac.lic` does this:

```
    scripts_to_unpause = []
    scripts_to_unpause = DRC.smart_pause_all
    ...
    DRC.unpause_all_list(scripts_to_unpause)
```

SekmehtDR commented:
Thanks, for sure worth a refresh/modernization on it. Will review and update with a new commit.

asechrest commented on tessera.lic:
```diff
@@ -0,0 +1,62 @@
+custom_require.call(%w[common common-items])
+
+class Tessera
+  include DRC
+  include DRCI
```
One thing we're trying to do with scripts moving forward is avoid the include lines in favor of the explicit prefix before methods.The primary reason is that it makes the code much more approachable for new contributors. It makes it easy to see where our helper methods live, and easy to go and reference them.

For, for example, you'd remove the `include DRC` line in favor of calling bput by using `DRC.bput` -- and any other methods that live there. Note: the `custom_require` line is still required.

SekmehtDR commented:
Thanks for all of the input guys, I'm doing some testing on my side with some of the newer almanac code, a lot of this makes sense especially around the pausing of scripts. I'll keep the DEV going on this, for now though I received DMs from one person saying it works great.

asechrest commented:
> Thanks for all of the input guys, I'm doing some testing on my side with some of the newer almanac code, a lot of this makes sense especially around the pausing of scripts. I'll keep the DEV going on this, for now though I received DMs from one person saying it works great.

Great work so far.

SekmehtDR commented:
I might need some help to fully modernize this with the smart pause stuff if someone wouldn't mind helping me. Otherwise the logic and everything is pretty solid. One person has said they enjoy being locked in trading all day long.

Once this becomes a thing, then we can bugfix/tweak it as needed.

SekmehtDR commented:
having someone test this baseline, with a startup delay timer, tester noted that scripts like bootstrap...etc. were causing issues at startup.

SekmehtDR commented:
I have the modernized script completed, testing now. Once its stable, I'll commit my changes and be ready for review upon letting you all know. Thanks.

rpherbig commented on tessera.lic:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  def use_tessera
+    settings = get_settings
```
We should try not to `get_settings` a second time (especially since `tessera_retry_interval` is already retrieved in `initialize`). Unless there is a reason to expect `tessera_retry_interval` will be different the second time?

rpherbig commented on tessera.lic:
```diff
@@ -0,0 +1,74 @@
+
+  def initialize
+    settings = get_settings
+    tessera_retry_interval = settings.tessera_retry_interval
```
Related to the other comment, I suggest making this an instance var (`@tessera_retry_interval`).

rpherbig commented on tessera.lic:
```diff
@@ -0,0 +1,74 @@
+    @tessera = settings.tessera_noun
+
+    DRC.message("Tessera Startup Options:")
+    DRC.message("[tessera_noun: #{@tessera}]\n[tessera_startup_delay: #{startup_delay} (seconds)]\n[tessera_retry_interval: #{tessera_retry_interval} (seconds)]\n[tessera last attempt: #{UserVars.tessera_last_use}]\n[tessera_no_use_scripts: #{@no_use_scripts}]")
```
I learned about this neat feature of Ruby recently and wanted to mention it. I'm not saying you have to use it, but it may make the code cleaner.

```
DRC.message(
	<<~STARTUP
      tessera_noun: #{@tessera}
      tessera_startup_delay: #{startup_delay} (seconds)
      tessera_retry_interval: #{tessera_retry_interval} (seconds)
      tessera last attempt: #{UserVars.tessera_last_use}
      tessera_no_use_scripts: #{@no_use_scripts}
    STARTUP
    )
```

Note: "STARTUP" is not special, you can name it whatever you want

rpherbig commented on tessera.lic:
```diff
@@ -0,0 +1,74 @@
+
+    return if Time.now - UserVars.tessera_last_use < tessera_retry_interval
+    return if @no_use_scripts.any? { |name| Script.running?(name) }
+    return if XMLData.room_title.include? 'Carousel Chamber'
```
What do you think about extracting this to a "no use rooms" setting like these?

https://github.com/rpherbig/dr-scripts/blob/567483a6bd76c4f8b82800f8b04fe846b63ebed0/profiles/base.yaml#L249

https://github.com/rpherbig/dr-scripts/blob/567483a6bd76c4f8b82800f8b04fe846b63ebed0/profiles/base.yaml#L1884

rpherbig commented on tessera.lic:
```diff
@@ -0,0 +1,74 @@
+
+    pause 1 until DRC.pause_all
+    waitrt?
+    DRC.bput('retreat', 'You get', 'You retreat', 'You are already')
```
`DRC.retreat` would be good here

SekmehtDR replied:
Thanks for the suggestion about DRC.retreat, if there's some standardized way to handle retreats it should be used.

rpherbig commented on profiles/base.yaml:
```diff
@@ -1857,6 +1857,43 @@ almanac_no_use_scripts:
 - mech-lore
 - corn-maze
 - burgle
+- safe-room
```
Was it intentional to add `safe-room` to `almanac_no_use_scripts` in a PR about tessera?

Should `safe-room` also or instead be in the list for tessera?

SekmehtDR replied:
`  def retreat(ignored_npcs = [])
    return if (DRRoom.npcs - ignored_npcs).empty?

    escape_messages = [
      /You are already as far away as you can get/,
      /You retreat from combat/,
      /You sneak back out of combat/,
      /Retreat to where/,
      /There's no place to retreat to/
    ]

    retreat_messages = [
      /retreat/,
      /sneak/,
      /grip on you/,
      /grip remains solid/,
      /You try to back/,
      /You must stand first/,
      /You stop advancing/,
      /You are already/
    ]

    loop do
      case DRC.bput("retreat", *escape_messages, *retreat_messages)
      when *escape_messages
        return true
      else
        DRC.fix_standing
      end
    end
  end`

So for me the fix standing part is the issue, because people lie down for healing. So this stands you up, then safe-room puts you down again and that cycle repeats and if there is a lot of people you are going to waiting a lot longer to be healed.

So needless to say since safe-room is inside, you aren't accumulating starlight which might make your tessera ask fail anyways, its probably better to just leave safe-room in there :)

I could put other logic around it to get it work, but this is the easiest way I can think of to handle it :)

SekmehtDR replied:
whoops, I see what you were saying, safe-room was accidently added to the almanac section, removed that.

SekmehtDR commented:
Okay, I have my final commit ready for this build. I've taken the suggestions (highly appreciated) and will include them for further approval. I'm running some testing on it now.

SekmehtDR commented:
Okies, this is ready to go :)

MahtraDR commented:
@SekmehtDR Sorry mate, I've had a super busy week, I haven't even logged into my trader.

I will test this tomorrow, and report back.

In the meantime, could you please resolve any open/pending comment as appropriate?

SekmehtDR commented:
any luck with testing @MahtraDR ? Let me know if I can assist with anything.

MahtraDR commented:
Hey mate, I'm sorry, life's been crazy and I hadn't logged that character in.

Looks good, and works. The only thing that gave me pause is this
```
>;tess                                                                                                                                                                                                                                       |
--- Lich: tessera active.                                                                                                                                                                                                                    |
Tessera Startup Options:                                                                                                                                                                                                                     |
                                                                                                                                                                                                                                             |
tessera_noun: tessera                                                                                                                                                                                                                        |
tessera_startup_delay: 15 (seconds)                                                                                                                                                                                                          |
tessera_retry_interval: 375 (seconds)                                                                                                                                                                                                        |
tessera last attempt: 2022-03-08 00:46:18 +0000                                                                                                                                                                                              |
tessera_no_use_scripts: ["sew", "carve", "tinker", "forge", "remedy", "shape", "enchant", "go2", "get2", "bescort", "pick", "locksmithing", "clean-leather", "clean-lumber", "craft", "athletics", "corn-maze", "burgle", "safe-room"]       |
tessera_no_use_rooms: [/\b(Carousel|Vault|Item Registration Office|Registrar's Office)\b/]        

```

MahtraDR commented:
Hey mate, I'm sorry, life's been crazy and I hadn't logged that character in.

Looks good, and works. The only thing that gave me pause is this
```
>;tess                                                                                                                                                                                                                                       |
--- Lich: tessera active.                                                                                                                                                                                                                    |
Tessera Startup Options:                                                                                                                                                                                                                     |
                                                                                                                                                                                                                                             |
tessera_noun: tessera                                                                                                                                                                                                                        |
tessera_startup_delay: 15 (seconds)                                                                                                                                                                                                          |
tessera_retry_interval: 375 (seconds)                                                                                                                                                                                                        |
tessera last attempt: 2022-03-08 00:46:18 +0000                                                                                                                                                                                              |
tessera_no_use_scripts: ["sew", "carve", "tinker", "forge", "remedy", "shape", "enchant", "go2", "get2", "bescort", "pick", "locksmithing", "clean-leather", "clean-lumber", "craft", "athletics", "corn-maze", "burgle", "safe-room"]       |
tessera_no_use_rooms: [/\b(Carousel|Vault|Item Registration Office|Registrar's Office)\b/]        

```

SekmehtDR commented:
No worries! I'll just remove it or now.

SekmehtDR commented:
> 

I just removed it from the script since it's not like the others. It really could be considered a debug type thing. But I added the almanac improved handling into my script. Should be good to go.
